### ISSUE TITLE
scVI training does not play nicely with other torch code

### ISSUE THREAD

**Issue by aksarkar** [YosefLab/scVI#214]
Module `scvi.inference.trainer` calls `torch.set_grad_enabled(False)` on import (line 16).

This silently breaks any code which builds torch graphs after `import scvi.inference`.

Is there a way to rewrite this with finer granularity?

**Comment by aksarkar** [comment]
I would recommend annotating functions with `@torch.no_grad` instead

**Comment by romain-lopez** [comment]
It looks like the current version of the code is half-annotated. I can see some with torch.autograd(True) and some with torch.autograd(False) but not everywhere. 

If I understand well, the only file that should be annotated with these decorators (@torch.no_grad) are the inference files ? In particular, we should add the decorator to any downstream analysis task (imputation, compute latent space, DE etc...)

@aksarkar @jeff-regier

**Comment by jeff-regier** [comment]
The training loops are the only places we need gradients. I'm not sure exactly how that translates to adding `torch.no_grad` annotations.

**Comment by romain-lopez** [comment]
I annotated the functions and deleted the line mentionned in this issue but it turns out the simple code still does not compute gradients after the following code:

`from scvi.inference import UnsupervisedTrainer`

Testing code:
`import torch
x = torch.ones(2, 2, requires_grad=True)
print(x)

y = x + 2
print(y)`

Still working on that

**Comment by romain-lopez** [comment]
OK I have solved the issue and ran the test + the notebooks. PR coming
